### PR TITLE
Noita: Update to use new Options API

### PR DESCRIPTION
**Pull request by ScipioWright**
## What is this fixing or adding?
Reworking the options to make it work with the new options API.
Also reworked stuff in several spots to use world: NoitaWorld instead of multiworld: MultiWorld

## How was this tested?
Generated a few games with different settings, connected it to the mod once

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

File: worlds/noita/__init__.py
Comment (heinermann) [inline worlds/noita/__init__.py:41]: What's up with this?

Comment (ScipioWright) [reply worlds/noita/__init__.py:41]: A mistake, woops

File: worlds/noita/__init__.py
Comment (alwaysintreble) [inline worlds/noita/__init__.py:43]:
```suggestion
        return self.options.as_dict("death_link", "bad_effects", etc)
```
do you really need the start inventory in your slot data? please don't send data you don't need.

Comment (ScipioWright) [reply worlds/noita/__init__.py]: Removed that and bad effects, forgot that start_inventory_from_pool was handled with item sends rather than something the client needs, been a long day

File: worlds/noita/Rules.py
Comment (ScipioWright) [reply worlds/noita/Rules.py:74]: Unfortunately, `forbid_items_for_player` takes a `Set[str]`, so we'd have to instead convert everything into sets. I'll leave it up to you, I don't really mind either way.

Comment (heinermann) [reply worlds/noita/Rules.py:74]: In that case, `items_hidden_from_shops` can be made into a set naturally, and other stuff can just use the set constructor `wands_to_forbid = set(wand_tiers[whatever])`

**Comment by heinermann** [comment]
You should be able to delete the `add_location` function in `Regions.py`.

File: worlds/noita/Events.py
Comment (alwaysintreble) [inline worlds/noita/Events.py:22]: why do you give typing to world everywhere else but here?

File: worlds/noita/Items.py
Comment (alwaysintreble) [inline worlds/noita/Items.py:54]:
```suggestion
    if not world.options.bad_effects:
```

File: worlds/noita/Regions.py
Comment (alwaysintreble) [inline worlds/noita/Regions.py:20]: i really think this reads better without these extra variable definitions. i also don't see anywhere that the `.value` is actually necessary.

File: worlds/noita/Regions.py
Comment (alwaysintreble) [inline worlds/noita/Regions.py:36]: other than here ig

File: worlds/noita/Regions.py
Comment (alwaysintreble) [inline worlds/noita/Regions.py:43]: i really don't think it's good practice to have multiple nested function calls that use the exact same name.

File: worlds/noita/Regions.py
Comment (alwaysintreble) [inline worlds/noita/Regions.py:47]: why does this function even exist?

Comment (alwaysintreble) [reply worlds/noita/Regions.py:47]: can't mark it, but same comment for create_item in Items.py line 25. that could just be the definition in the world and then all you have to do is `world.create_item(item_name)`

File: worlds/noita/Rules.py
Comment (alwaysintreble) [inline worlds/noita/Rules.py:74]: I think it'd make more sense to either just directly pass the location to this method or remove this function all together personally, but this one isn't that big of a deal. I understand the idea of wanting to segment everything apart, but you have everything so segmented that it makes it more difficult to follow what's actually happening.

File: worlds/noita/__init__.py
Comment (alwaysintreble) [inline worlds/noita/__init__.py:52]: I wouldn't deny based on this, but I really don't like all these nested calls just for the sake of it. it's hard to follow, and every function created takes up more memory and time.

**Review by alwaysintreble — CHANGES_REQUESTED** [review]

File: worlds/noita/Regions.py
Comment (heinermann) [reply worlds/noita/Regions.py:43]: `add_locations` can be renamed to `create_locations`.

File: worlds/noita/Regions.py
Comment (heinermann) [reply worlds/noita/Regions.py:47]: > why does this function even exist?

To make `create_all_regions_and_connections` readable and keep the world class slim.

File: worlds/noita/Rules.py
Comment (heinermann) [reply worlds/noita/Rules.py:74]: The location needs to be retrieved in every case which is what this is really facilitating. If anything more logic could be moved into this function to handle multiple locations, then code at three of the callers can be improved such as.

```py
def ban_items_from_shops(multiworld: MultiWorld, player: int) -> None:
    for location_name in Locations.location_name_to_id.keys():
        if "Shop Item" in location_name:
            forbid_items_at_location(multiworld, location_name, items_hidden_from_shops, player)
```

could be

```py
def ban_items_from_shops(multiworld: MultiWorld, player: int) -> None:
    shop_locations = [name for name in Locations.location_name_to_id.keys() if "Shop Item" in name]
    forbid_items_at_locations(multiworld, shop_locations, items_hidden_from_shops, player)
```

The two for loops for the other callers would be removed.

**Comment by ScipioWright** [comment]
Converting to draft until I have the time to fix the merge conflicts and deal with the comments

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**
